### PR TITLE
Legacy RPG Gunnery Skills

### PR DESCRIPTION
**Pull request by sagnam**
Implement RPG gunnery skills for all gunnery types.  Fix skills when loading campaign with RPG skills enabled but person only has the non-RPG gunnery skill only for that type.

### CODE REVIEW

**Comment by sagnam** [comment]
Ah!  This depends on MM.  Not sure how that dependency is handled on the remote.

**Comment by sixlettervariables** [comment]
If your branch names match per our build instructions, the CI actions will work. (LGTM will not)

File: MekHQ/src/mekhq/gui/view/PersonViewPanel.java [hunk 1]
```diff
@@ -1167,17 +1168,27 @@ private JPanel fillSkills() {
         double weight = 0.5;
 
         int j = 0;
-        for(int i = 0; i < SkillType.getSkillList().length; i++) {
-            if(person.hasSkill(SkillType.getSkillList()[i])) {
+        for(String skill : SkillType.getSkillList()) {
+            if(person.hasSkill(skill)) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/view/PersonViewPanel.java:1172]:
```suggestion
        for (String skill : SkillType.getSkillList()) {
            if (person.hasSkill(skill)) {
```

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 2]
```diff
@@ -3727,10 +3734,41 @@ private void calcCompositeCrew() {
             }
 
             if (entity instanceof Tank
-                    && Compute.getFullCrewSize(entity) == 1
-                    && p.hasSkill(gunType)) {
-                sumGunnery += p.getSkill(gunType).getFinalSkillValue();
-                nGunners++;
+                    && Compute.getFullCrewSize(entity) == 1) {
+                if (SkillType.isRPGGunnery(gunType) &&
+                        campaign.getGameOptions().booleanOption(OptionsConstants.RPG_RPG_GUNNERY)) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:3739]: These should be in the opposite order, always check if you are using the RPG gunnery option first.

File: MekHQ/src/mekhq/campaign/unit/Unit.java [hunk 3]
```diff
@@ -3741,10 +3779,42 @@ private void calcCompositeCrew() {
             if(p.getHits() > 0 && !usesSoloPilot()) {
                 continue;
             }
-            if(p.hasSkill(gunType)) {
-                sumGunnery += p.getSkill(gunType).getFinalSkillValue();
-                nGunners++;
+
+            if (SkillType.isRPGGunnery(gunType) &&
+                    campaign.getGameOptions().booleanOption(OptionsConstants.RPG_RPG_GUNNERY)) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/Unit.java:3784]: These should be in the opposite order, always check if you are using the RPG gunnery option first.

File: MekHQ/src/mekhq/gui/view/PersonViewPanel.java [hunk 4]
```diff
@@ -1167,17 +1168,27 @@ private JPanel fillSkills() {
...
                     firsty = 0;
                     weight = 1.0;
                 }
                 lblName = new JLabel(
-                    String.format(resourceMap.getString("format.itemHeader"), SkillType.getSkillList()[i])); //$NON-NLS-1$
-                lblValue = new JLabel(person.getSkill(SkillType.getSkillList()[i]).toString());
+                    String.format(resourceMap.getString("format.itemHeader"), skill)); //$NON-NLS-1$
+                lblValue = new JLabel(person.getSkill(skill).toString());
                 gridBagConstraints = new GridBagConstraints();
                 gridBagConstraints.gridx = addition;
```
[the remark this answers is not in the document]
Comment (sagnam) [reply MekHQ/src/mekhq/gui/view/PersonViewPanel.java:1193]: The String.format?  It's a continuation from the previous line
`                lblName = new JLabel(
                    String.format(resourceMap.getString("format.itemHeader"), skill)); //$NON-NLS-1$
                lblValue = new JLabel(person.getSkill(skill).toString());
`

[the remark this answers is not in the document]
Comment (sagnam) [reply MekHQ/src/mekhq/gui/view/PersonViewPanel.java:1193]: Interestingly the code block formatting removed the newline.

File: MekHQ/src/mekhq/gui/dialog/CustomizePersonDialog.java [hunk 5]
```diff
@@ -1056,16 +1066,18 @@ public void refreshSkills() {
...
+                if (skillChks.get(type).isSelected()) {
+                    int lvl = (Integer) skillLvls.get(type).getModel().getValue();
+                    int b = (Integer) skillBonus.get(type).getModel().getValue();
+                    person.addSkill(type, lvl, b);
+                } else {
+                    person.removeSkill(type);
+                }
             }
-        }
+    	}
```
[the remark this answers is not in the document]
Comment (sagnam) [reply MekHQ/src/mekhq/gui/dialog/CustomizePersonDialog.java]: There is an additional check added that resulted in another level of indentation.  Perhaps that's what is causing it to look off?

**Comment by sagnam** [comment]
> Partial review, but I've also found a few issues while testing it.

> Infantry Skills in the Personnel Tab has a NPE when selected with this branch.
Do you have a campaign I can use to recreate?  In my test campaign I'm not seeing this but I'm probably missing something.

> XP costs seem to be a constant 8 for the new skills.
Huh, you are right.  I didn't notice this before.  I'll dig in to see where this is calculated.  Looking at an existing save I have (before this change), the single Gunnery/Mech cost option was 16,8,8,8,8,8,8,8,-1,-1,-1 in the save which is the same as these new skills.  The skill advancement rules for 3ed RPG rules are vastly different that what's in MekHQ currently anyway, so I have to assume anyone using this option will have to track XP separately.  I'm not sure how to handle it to be honest.  Seems like the easiest would just be to keep it as is and allow users to spend XP on the individual skills as they choose.  This would be an unofficial system though, so maybe RPG gunnery should be labeled as unofficial instead of Legacy.

> There should be a MekHQ-side option to enable/disable it (see Tech Level for an example).
In addition to the MegaMek option, or instead of?

> When converting between the two types it does not apply the changes based on averaging the skill level.
Do you think this is worthwhile?  If the person doesn't have the relevant skill this makes sense for sure.  If they do have the skill checked, even if it's not currently visible, I worry about overwriting the correct value.  For example, having 2/3/4 then switching to non-rpg would take the average of 3.   When switching back to RPG we would replace the individual values with the average of 3/3/3.  Seems like switching the first time should calculate the average of 3, and witching back should retain the 2/3/4.

> It should only need to track one form of skills, either the Legacy RPG style or the current MekHQ style. Both is a bit superfluous and will clutter up the save files. (converting between the two whenever the option is changed in MekHQ).
This makes sense.  I was worried about the above scenario when saving, but maybe I can just save the selected format to disk and keep the other in memory in case they switch back.  When they load they would lose the one stored in memory and get the averaged value if they toggle the option.  Should I also filter the customizations from the Skills tab of the Campaign Options?

**Comment by sagnam** [comment]
Updated campaign loading and also monitor changes to options in the campaign.  When RPG gunnery skills setting is enabled, update all personnel to add RPG gunnery skills if those skills aren't already populated for a person that has the associated single gunnery skill and defaulting the value of those RPG skills to that of the associated single gunnery skill.  When RPG gunnery skills setting is disabled, update personnel that don't have a single gunnery skill for the associated RPG gunnery skills by calculating the average of the 3 and assigning that value to the single skill.

**Comment by Windchild292** [comment]
After discussion and multiple internal reviews the dev team has decided this adds too much technical debt to merge. The skill implementation needs a complete rewrite, and with that it will preferably handle this setup without the increased difficulty.

**Pull request closed without merging**